src/main/java/frc/robot/subsystems/drivetrain/SwerveDrive.java

```diff
@@ -0,0 +1,231 @@
+            swerveModules[k].setAngle(controls[k][1]);
+        }
+
+        double sumx = 0;
+        double sumy = 0;
+        for (int j = 0; j < 4; j++) {
+            sumx += velocities[j * 2];
+            sumy += velocities[j * 2 + 1];
+        }
+        double[] target = Utils.cartesianToPolar(sumx, sumy);
+        FireLog.log("target velocity", target[0]);
+        FireLog.log("target angle", target[1]);
+    }
+
+    /**
+     * turns the joystick inputs into the robot heading
+     *
+     * @param forward    the Y value of the joystick
+     * @param strafe     the X value of the joystick
+     * @param rotation   the rotation Z of the joystick
+     * @param robotAngle the current angle of the robot in radians
+     * @return an array of the robot heading
+     */
+    public static double[] getRobotHeading(double forward, double strafe, double rotation, double robotAngle) {
+        // turns the joystick values into the heading of the robot
+        forward *= Constants.SwerveDrive.SPEED_MULTIPLIER;
+        strafe *= Constants.SwerveDrive.SPEED_MULTIPLIER;
+        rotation *= Constants.SwerveDrive.ROTATION_MULTIPLIER;
+        // multiplies the 2D rotation matrix by the robot heading, there by rotating the coordinate system
+        // see https://en.wikipedia.org/wiki/Rotation_matrix
+        double[][] rotationMat = {{Math.cos(robotAngle), -Math.sin(robotAngle)},
+                {Math.sin(robotAngle), Math.cos(robotAngle)}};
+        double[] speeds = Utils.matrixVectorMult(rotationMat, new double[]{forward, strafe});
+        System.out.println("forward" + forward);
+        System.out.println("strafe" + strafe);
+
+        // if the drive style is field oriented, changes the forward and strafe to be according to the field coordinate system
+        if (isFieldOriented) {
+            forward = speeds[0];
+            strafe = speeds[1];
+        }
+        System.out.println("relative forward" + speeds[0]);
+        System.out.println("relative strafe" + speeds[1]);
+
+        double[] robotHeading = {forward, strafe, rotation};
+
+        return robotHeading;
+    }
+
+    /**
+     * calculates the velocity vector of each wheel
+     *
+     * @param robotHeading the three joystick outputs:
+     *                     forward the heading of the robot in the Y direction
+     *                     strafe the heading of the robot in the X direction
+     *                     rotation the rotation of the robot
+     * @return an array of length 8 in which each pair is the X and Y velocities of each wheel
+     */
+    public static double[] calculateWheelVelocities(double[] robotHeading) {
+        // multiplies M by the robotHeading to obtain the wheel velocities
+        double[] wheelVelocities = Utils.matrixVectorMult(dynamics, robotHeading);
+        return wheelVelocities;
+    }
+
+    /**
+     * set the angle of the wheels on the robot to lock the robot in place
```

Comment:
Can you explain what that means?

Reply:
Lock angles are used for when there is defence on the robot, the wheels point outwards and it is nearly inmpossible to move the robot.